pkg/web/metricsHandler_test.go

```diff
@@ -0,0 +1,140 @@
+package web
+
+import (
+	"testing"
+
+	"encoding/json"
+	"io/ioutil"
+	"net/http"
+	"net/http/httptest"
+	"github.com/aerogear/aerogear-metrics-api/pkg/mobile"
+	"bytes"
+	"github.com/stretchr/testify/mock"
+	"github.com/stretchr/testify/assert"
+	"github.com/pkg/errors"
+)
+
+type MockMetricsService struct {
+	mock.Mock
+}
+
+func (m *MockMetricsService) Create(metric mobile.Metric) (mobile.Metric, error) {
+	args := m.Called(metric)
+	return args.Get(0).(mobile.Metric), args.Error(1)
+}
+
+func setupMetricsHandler(service MetricsServiceInterface) *httptest.Server {
+	router := NewRouter()
+	metricsHandler := NewMetricsHandler(service)
+	MetricsRoute(router, metricsHandler)
+	return httptest.NewServer(router)
+}
+
+func TestMetricsEndpointShouldPassReceivedDataToMetricsService(t *testing.T) {
+	metric := mobile.Metric{
+		ClientTimestamp: 1234,
+		ClientId:        "client123",
+		App: mobile.AppMetric{
+			ID:         "deadbeef",
+			SDKVersion: "1.2.3",
+			AppVersion: "27",
+		},
+		Device: mobile.DeviceMetric{
+			Platform:        "android",
+			PlatformVersion: "19",
+		},
+	}
+
+	byteBuffer := new(bytes.Buffer)
+	err := json.NewEncoder(byteBuffer).Encode(metric)
+	assert.Nil(t, err, "did not expect an error marshaling metric")
+
+	mockMetricsService := new(MockMetricsService)
+	mockMetricsService.On("Create", metric).Return(metric, nil)
+
+	s := setupMetricsHandler(mockMetricsService)
+	defer s.Close()
+
+	res, err := http.Post(s.URL+"/metrics", "application/json", byteBuffer)
+	assert.Nil(t, err, "did not expect an error posting metrics")
+
+	defer res.Body.Close()
+	_, err = ioutil.ReadAll(res.Body)
+
+	assert.Equal(t, 200, res.StatusCode)
+
+	mockMetricsService.AssertExpectations(t)
+}
+
+func TestMetricsEndpointShouldReturn500WhenThereIsAnErrorInMetricsService(t *testing.T) {
+	metric := mobile.Metric{
+		ClientTimestamp: 1234,
+		ClientId:        "client123",
+		App: mobile.AppMetric{
+			ID:         "deadbeef",
+			SDKVersion: "1.2.3",
+			AppVersion: "27",
+		},
+		Device: mobile.DeviceMetric{
+			Platform:        "android",
+			PlatformVersion: "19",
+		},
+	}
+
+	byteBuffer := new(bytes.Buffer)
+	err := json.NewEncoder(byteBuffer).Encode(metric)
+	assert.Nil(t, err, "did not expect an error marshaling metric")
+
+	mockMetricsService := new(MockMetricsService)
+	mockMetricsService.On("Create", metric).Return(mobile.Metric{}, errors.New("Metrics service error!")) // mock the service so that it returns an error
```

Comment:
I had to import "github.com/pkg/errors" just because of this line.
What I want is making the mock service return an error on the call for this test case.
I tried
```
mockMetricsService.On("Create", metric).Return(mobile.Metric{}, err.Error())
```
but it causes the execution stop on this line.

@maleck13 @darahayes @wtrocki any other way of doing this?

Reply:
you can do ```fmt.Errorf()``` if that helps. Not entirely understanding the issue though. There is also the built in errors package, change the import to be ```import errors```

Reply:
Yes please just use the built in errors package. It works pretty much the same way with `errors.New("some message")`